### ISSUE TITLE
Fail to load SDL3

### ISSUE THREAD

**Issue by gruelingpine185** [libsdl-org/SDL#9017]
After building SDL3 from source and installing it with the instructions according to [INSTALL.md](https://github.com/libsdl-org/SDL/blob/main/INSTALL.md), I wrote some code to create a window in the form:
```c
#include <SDL3/SDL.h>

#include "game.h"


int main(void) {
    umbra_game game = {0};
    if(!umbra_create_game(&game)) return 1;

    while(game.state != UMBRA_QUIT_STATE) {
        umbra_render_game(&game);
        umbra_game_handle_events(&game);
        SDL_Delay(1000 / 60);
    }

    umbra_destroy_game(&game);
    return 0;
}
```

The implementation details of the `umbra_*`. functions don't matter though. As long as I use functions from libSDL3.dylib, link it,

```sh
clang -Iinc src/*.c -o umbra $(pkg-config --static --libs sdl3)
```

and try to run the resulting program, I get a message like this:
```sh
game main +? ➜ ./umbra
dyld[75562]: Library not loaded: @rpath/libSDL3.1.0.0.dylib
  Referenced from: <FCBC77A5-4C67-3795-876C-F8B92D87B069> /Users/gruelingpine/projects/game/umbra
  Reason: no LC_RPATH's found
zsh: abort      ./umbra
```
with an exit code of 134.

__Platform__ MacOS 14.3 (Sonoma)
__SDL Version__ 3.1+ (submodule from master branch)

**Comment by gruelingpine185** [comment]
Some searches on google suggested running the otool command. Here is the result of that:
```sh
game main +? ➜ otool -L umbra
umbra:
	@rpath/libSDL3.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreMedia.framework/Versions/A/CoreMedia (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo (compatibility version 1.2.0, current version 1.5.0)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 24.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/ForceFeedback.framework/Versions/A/ForceFeedback (compatibility version 1.0.0, current version 1.0.2)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 170.0.0)
	/System/Library/Frameworks/CoreAudio.framework/Versions/A/CoreAudio (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AudioToolbox.framework/Versions/A/AudioToolbox (compatibility version 1.0.0, current version 1000.0.0)
	/System/Library/Frameworks/AVFoundation.framework/Versions/A/AVFoundation (compatibility version 1.0.0, current version 2.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 2202.0.0)
	/System/Library/Frameworks/GameController.framework/Versions/A/GameController (compatibility version 1.0.0, current version 11.2.3, weak)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 341.35.0, weak)
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0, weak)
	/System/Library/Frameworks/CoreHaptics.framework/Versions/A/CoreHaptics (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
```

**Comment by slouken** [comment]
macOS is a little weird, because how you link depends on whether or not you're using frameworks and the app bundle format. Are you building an app bundle or a single executable?

**Comment by gruelingpine185** [comment]
I'm building a single Executable.

**Comment by gruelingpine185** [comment]
The only Frameworks that I use are whatever SDL uses under the hood. This appears to be:
```sh
game main +? ➜ pkg-config --static --libs sdl3
-L/usr/local/lib/pkgconfig/../../lib -lSDL3 -Wl,-framework,CoreMedia -Wl,-framework,CoreVideo -Wl,-framework,Cocoa -Wl,-framework,IOKit -Wl,-framework,ForceFeedback -Wl,-framework,Carbon -Wl,-framework,CoreAudio -Wl,-framework,AudioToolbox -Wl,-framework,AVFoundation -Wl,-framework,Foundation -Wl,-weak_framework,GameController -Wl,-weak_framework,Metal -Wl,-weak_framework,QuartzCore -Wl,-framework,CoreHaptics -lm -liconv
```

**Comment by slouken** [comment]
Does it work if you move libSDL3.1.0.0.dylib into the same directory as your executable?

**Comment by slouken** [comment]
FYI, if you ever ship on macOS you'll want to switch to creating an app bundle and using Xcode to build SDL as a framework.

**Comment by gruelingpine185** [comment]
That's good to know. Once I get to that point I'll figure out the steps I need to take to get that done. As for moving the library into the same directory as the executable, it still doesn't work. This was for the shared library. I'm not sure what `libSDL3_test.a` is for.

**Comment by madebr** [comment]
Side questions:
- what is the use of `--static` here? You're still linking to a shared SDL3 dylib, not a static SDL3 library.
- there is no Apple-equivalent of `-Wl,-rpath=...`?

Also, CMake can build a SDL framework by configuring with `-DSDL_FRAMEWORK=ON`.

`libSDL3_test.a` contains utility functions for the SDL tests. It's used by the satellite libraries (SDL_image, ...).
Normal applications usually don't have a need for it.

**Comment by gruelingpine185** [comment]
Linking statically was the intent. I didn't realize that SDL3 didn't have a static library until you asked me to move the libraries next to my executable. I can try using `--shared`, but as pointed out with the output from otool: `libSDL3.1.0.0.dylib` and within the output from pkg-config: `-lSDL3`, I don't think it makes a difference since they both are pointing to the shared library

**Comment by gruelingpine185** [comment]
> there is no Apple-equivalent of -Wl,-rpath=...?

I have no idea. Appending `-Wl,rpath=.` results in
```sh
game main +? ➜ clang -Iinc src/*.c -o umbra `pkg-config --static --libs sdl3` -Wl,rpath=.
ld: library 'SDL3' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

If `-Wl,-rpath=` exists, it is not in the (clang manual)[https://clang.llvm.org/docs/UsersManual.html].

**Comment by gruelingpine185** [comment]
If I use pkg-config for shared library flags I get this:
```sh
game main +? ➜ pkg-config  --libs sdl3
-L/usr/local/lib/pkgconfig/../../lib -lSDL3
game main +? ➜ clang -Iinc src/*.c -o umbra -L. -lSDL3
ld: library 'SDL3' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
game main +? ➜ ls
inc		libSDL3.dylib	src		vendor
```

**Comment by madebr** [comment]
Try
```
clang -L/usr/local/lib -lSDL3 -Wl,-rpath=/usr/local/lib -Iinc src/*.c -o umbra
```

**Comment by gruelingpine185** [comment]
> clang -L/usr/local/lib -lSDL3 -Wl,-rpath=/usr/local/lib -Iinc src/*.c -o umbra

Resulted in ```sh
ld: unknown options: -rpath=/usr/local/lib
```

**Comment by gruelingpine185** [comment]
I did find a program named `install_name_tool`.
<img width="1439" alt="Screenshot 2024-02-07 at 2 13 46 PM" src="https://github.com/libsdl-org/SDL/assets/66073333/c0da9ed9-cc49-42d5-adf8-dfe5c95622e9">

I'll figure out how to use it and see if I can change the path with it.

**Comment by madebr** [comment]
There needed to be a comma after `-rpath`.
```
clang -L/usr/local/lib -lSDL3 -Wl,-rpath,/usr/local/lib -Iinc src/*.c -o umbra
```

**Comment by gruelingpine185** [comment]
Thanks :)

`-L/usr/local/lib` is already in the compiler's search path. It can be omitted.

For the record, this command below is what resolved the "Library not loaded" error and allowed for me to run the program.
```sh
clang -lSDL3 -Wl,-rpath,/usr/local/lib -Iinc src/*.c -o <executable>
```

I did receive an exit code of 1 when running the program though. I'll debug it and will report back if I think it is related to this issue. Thanks for your help. It was greatly appreciated :)

**Comment by xesf** [comment]
I had a similar issue and I got it working with the merged details #9147 
`-L$(BIN) -Wl,-rpath,${BIN} -lSDL3`

Although is working as dynamic linking and not static.
